**Theory Questions**

1. Difference between a Function and a Method in Python

    - A function is an independent block of code that performs a specific task and can exist on its own.
    - A method is similar to a function but it is tied to an object and usually operates on data belonging to that object.

      

In [27]:

      def show_message():
          print("Hello from a function!")

      class Student:
        def greet(self):     # method
          print("Hello from a method!")

      show_message()
      s = Student()
      s.greet()

Hello from a function!
Hello from a method!


2. Function Arguments and Parameters

    - Parameters are names listed inside the function definition.

    - Arguments are actual values you pass to those parameters when calling the function.

    - Think of parameters as “empty boxes” and arguments as “what you put inside the boxes.”

      Example:


In [26]:
def multiply(a, b):   # a, b are parameters
    return a * b

result = multiply(4, 6)  # 4, 6 are arguments


3. Different Ways to Define and Call a Function

    There are multiple ways:

    Using def — the standard way.

    Using lambda — for short one-line functions.

    Using recursion — when a function calls itself.

    Example:

In [24]:
# Normal function
def square(x):
    return x**2

# Lambda function
square2 = lambda x: x**2

print(square(5))
print(square2(5))


25
25


4. Purpose of the return Statement

    - The return statement ends a function and sends a result back to where the function was called.
    - Without return, a function gives back None by default.

    Example:

In [23]:
def average(a, b):
    return (a + b) / 2

print(average(10, 20))  # Output: 15.0


15.0


5. Iterators vs Iterables

    - Iterable: Any object you can loop over (like lists, tuples, or strings).

    - Iterator: An object that keeps track of the current position and produces elements one at a time using next().

In [22]:
data = [3, 6, 9]    # iterable
it = iter(data)     # iterator

print(next(it))  # 3
print(next(it))  # 6
print(next(it))  # 9


3
6
9


6. Generators in Python

    - A generator is a function that remembers its state between calls and yields one value at a time instead of returning all at once.
    - It is created using the yield keyword.

   Example:

In [21]:
def countdown(n):
    while n > 0:
        yield n
        n -= 1

for i in countdown(3):
    print(i)


3
2
1


7. Advantages of Generators

    - Saves Memory: Doesn’t store the whole sequence in memory.

    - Lazy Evaluation: Produces items only when needed.

    - Convenient for Infinite Sequences.

    - Example: Reading a large file line by line instead of loading the entire file at once.

8. Lambda Function

    - A lambda function is a small anonymous function written in one line using the lambda keyword.
    - They are useful when a short function is required temporarily.

  Example:

In [20]:
add = lambda x, y: x + y
print(add(10, 5))  # Output: 15


15


9. Purpose and Usage of map() Function

    - map() applies a specific function to every element of an iterable and returns a new iterable (a map object).

  Example:

In [19]:
numbers = [2, 4, 6, 8]
doubles = map(lambda x: x * 2, numbers)
print(list(doubles))  # [4, 8, 12, 16]


[4, 8, 12, 16]


10. Difference between map(), reduce(), and filter()

Function	What it does	Output Type	Example

  - map()
      - Applies a function to each element
      - Iterable
      - map(lambda x:x*3, [1,2,3]) → [3,6,9]

  - filter()
    - Keeps only elements meeting a condition
    - Iterable
    - filter(lambda x:x>3, [1,4,5]) → [4,5]

  - reduce()
    - Combines elements into one using a function
    - Single Value
    - reduce(lambda x,y:x+y,[1,2,3]) → 6

11. Internal Mechanism of reduce() (Sum Operation)

    List: [47, 11, 42, 13]

Code:

In [18]:
from functools import reduce
reduce(lambda x, y: x + y, [47, 11, 42, 13])


113

  - Step-by-step working:

  - Take first two → 47 + 11 = 58

  - Add next → 58 + 42 = 100

  - Add last → 100 + 13 = 113

**PRACTICAL QUESTIONS**

In [2]:
# 1. Write a Python function that takes a list of numbers as input and returns the sum of all even numbers in the list.
def sum_of_evens(numbers):
    even_sum = sum([num for num in numbers if num % 2 == 0])
    return even_sum

# Example
print(sum_of_evens([4, 7, 10, 15, 22]))  # Output: 36


36


In [3]:
# 2. Create a Python function that accepts a string and returns the reverse of that string.
def reverse_string(text):
    return text[::-1]

# Example
print(reverse_string("Python"))  # Output: nohtyP


nohtyP


In [4]:
# 3. Implement a Python function that takes a list of integers and returns a new list containing the squares of each number.
def list_of_squares(nums):
    return [n ** 2 for n in nums]

# Example
print(list_of_squares([2, 5, 7]))  # Output: [4, 25, 49]


[4, 25, 49]


In [5]:
# 4. Write a Python function that checks if a given number is prime or not from 1 to 200.
def check_prime(num):
    if num < 2:
        return False
    for i in range(2, int(num ** 0.5) + 1):
        if num % i == 0:
            return False
    return True

# Example
for n in range(1, 20):
    if check_prime(n):
        print(f"{n} is Prime")


2 is Prime
3 is Prime
5 is Prime
7 is Prime
11 is Prime
13 is Prime
17 is Prime
19 is Prime


In [6]:
# 5. Create an iterator class in Python that generates the Fibonacci sequence up to a specified number of terms.
class Fibonacci:
    def __init__(self, limit):
        self.limit = limit
        self.a, self.b = 0, 1
        self.count = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self.count < self.limit:
            value = self.a
            self.a, self.b = self.b, self.a + self.b
            self.count += 1
            return value
        else:
            raise StopIteration

# Example
for num in Fibonacci(8):
    print(num, end=" ")  # Output: 0 1 1 2 3 5 8 13


0 1 1 2 3 5 8 13 

In [7]:
# 6. Write a generator function in Python that yields the powers of 2 up to a given exponent.
def powers_of_two(n):
    for i in range(n + 1):
        yield 2 ** i

# Example
print(list(powers_of_two(6)))  # Output: [1, 2, 4, 8, 16, 32, 64]


[1, 2, 4, 8, 16, 32, 64]


In [17]:
import os

# 7. Implement a generator function that reads a file line by line and yields each line as a string.
def file_line_reader(filename):
    # If the file does not exist, create one with sample content
    if not os.path.exists(filename):
        with open(filename, 'w') as f:
            f.write("Python is fun\nGenerators are efficient\nThis file was auto-created\n")

    with open(filename, 'r') as file:
        for line in file:
            yield line.strip()

# Example
for line in file_line_reader('sample.txt'):
    print(line)


Python is fun
Generators are efficient
This file was auto-created


In [9]:
# 8. Use a lambda function in Python to sort a list of tuples based on the second element of each tuple.
data = [(1, 'banana'), (2, 'apple'), (3, 'cherry'), (4, 'date')]
sorted_data = sorted(data, key=lambda x: x[1])

print(sorted_data)


[(2, 'apple'), (1, 'banana'), (3, 'cherry'), (4, 'date')]


In [10]:
# 9. Write a Python program that uses map() to convert a list of temperatures from Celsius to Fahrenheit.
temps_c = [0, 25, 37, 100]
temps_f = list(map(lambda c: round((c * 9/5) + 32, 2), temps_c))

print(temps_f)


[32.0, 77.0, 98.6, 212.0]


In [11]:
# 10. Create a Python program that uses filter() to remove all the vowels from a given string.
def remove_vowels(text):
    vowels = 'aeiouAEIOU'
    return ''.join(filter(lambda ch: ch not in vowels, text))

# Example
print(remove_vowels("Artificial Intelligence"))


rtfcl ntllgnc


In [12]:
# 11. Imagine an accounting routine used in a book shop.
# It works on a list with sublists containing: [Order Number, Book Title, Quantity, Price per Item]
# Write a Python program using lambda and map which returns a list of 2-tuples.
# Each tuple contains the order number and the total product price.
# Add €10 if the total value is below €100.

orders = [
    [34587, "Learning Python, Mark Lutz", 4, 40.95],
    [98762, "Programming Python, Mark Lutz", 5, 56.80],
    [77226, "Head First Python, Paul Barry", 3, 32.95],
    [88112, "Einführung in Python3, Bernd Klein", 3, 24.99]
]

# Using map() and lambda for transformation
result = list(map(lambda order: (
    order[0],
    order[2] * order[3] if order[2] * order[3] > 100 else order[2] * order[3] + 10
), orders))

print(result)


[(34587, 163.8), (98762, 284.0), (77226, 108.85000000000001), (88112, 84.97)]
